File: packet/packet.py
#!/usr/bin/env python3
import sys
import struct 
MAGIC_CODE='*#$^'
if len(sys.argv) < 4:
    print('must have three paramater about <firmware_name> <firmware_version> <firmware_boardid>')
    sys.exit(127)
firmware_name=sys.argv[1]
firmware_version=sys.argv[2]
firmware_boardid=sys.argv[3]
SKU=1

def calculate_checksum(f_buffer):
    size=len(f_buffer)
    c0,c1=0,0
    for byte in f_buffer:
        c0 += int(byte)
        c1 += c0
    b = (c0 & 65535) + ((c0 >> 16) & 65535 )
    c0 = (( b >> 16) + b ) & 65535
    b = (c1 & 65535) + ((c1 >>16) & 65535)
    c1 = ((b>>16)+b) & 65535
    checksum= ((c1 << 16) | c0)
    return checksum
def validation_version(firmware_version):
    tmp_version = firmware_version.split('_',1)
    main_ver = tmp_version[0]
    sub_ver = tmp_version[1]
    maintmp_ver_list=main_ver.split ('.',3)
    sub_ver_list=sub_ver.split('.',2)
    main_ver_list= maintmp_ver_list + sub_ver_list
    main_ver_list = list (map(int , main_ver_list) )
    if len(main_ver_list) != 7:
        print ('Version must need x.x.x.x_x.x.x')
        sys.exit (127)
    for i in  range(len(main_ver_list)):
        if main_ver_list[i] > 255:
            print('each version number must lower than 255')
            sys.exit(127)
        print (main_ver_list[i])
    
    return bytearray(main_ver_list)



def main ():
    header_length=40 + len(firmware_boardid)
    versionlist = validation_version (firmware_version)
    rootfs_cksum=0
    rootfs_length=0
    chk_firmware_path = firmware_name + ".chk"
    with open(firmware_name,"rb") as f:
        binary_data=f.read()
    print (type(binary_data))
    f_buffer = bytearray(binary_data)
    kernel_cksum=calculate_checksum(f_buffer)
    kernel_length=len(f_buffer)
    image_cksum=kernel_cksum
    header_cksum =0 
    with open(chk_firmware_path,"wb") as f:
        f.write(bytearray(MAGIC_CODE, encoding='ASCII'))
        f.write(struct.pack('<I',header_length)[::-1])
        f.write(struct.pack('B',SKU))
        f.write (versionlist)
        f.write (struct.pack('<I',kernel_cksum)[::-1])
        f.write (struct.pack('<I',rootfs_cksum)[::-1])
        f.write (struct.pack('<I',kernel_length)[::-1])
        f.write (struct.pack('<I',rootfs_length)[::-1])
        f.write (struct.pack('<I',image_cksum)[::-1])
        f.write (struct.pack('<I',header_cksum)[::-1])
        f.write(bytearray(firmware_boardid, encoding='ASCII'))
    with open (chk_firmware_path,"rb") as f:
        header_buffer = f.read()


    header_buffer= bytearray(header_buffer)
    header_cksum= calculate_checksum(header_buffer)
    with open (chk_firmware_path ,"wb") as f:
        f.write(bytearray(MAGIC_CODE, encoding='ASCII'))
        f.write(struct.pack('<I',header_length)[::-1])
        f.write(struct.pack('B',SKU))
        f.write (versionlist)
        f.write (struct.pack('<I',kernel_cksum)[::-1])
        f.write (struct.pack('<I',rootfs_cksum)[::-1])
        f.write (struct.pack('<I',kernel_length)[::-1])
        f.write (struct.pack('<I',rootfs_length)[::-1])
        f.write (struct.pack('<I',image_cksum)[::-1])
        f.write (struct.pack('<I',header_cksum)[::-1])
        f.write(bytearray(firmware_boardid, encoding='ASCII'))
        f.write(f_buffer)

if __name__ == '__main__':
    main()

File: packet/packet.lua
#!/usr/bin/env lua
firmware_path= arg[1]
version = arg[2]
board_id = arg[3]
sku=1
kernel_cksum=0
rootfs_cksum=0
kernel_length=0
rootfs_length=0
image_cksum=0
header_cksum=0

function calculate_cksum(s,size)
	local c0 , c1 = 0 , 0
	for i=1,size do 
		local byte = string.byte(s:sub(i,i))
		c0 =  c0+ byte
		c1 = c1 +c0
	end
	b= ( c0 & 65535) + ((c0 >>16) & 65535)
	c0 = (( b >> 16 ) +b ) & 65535
	b= (c1 & 65535 ) + ((c1 >> 16) & 65535)
	c1= ((b >> 16 ) + b ) & 65535

	cksum = ((c1 << 16) | c0 )
	return cksum
end

function validation_version(version ) 
	local t = {}
	local main_ver=""
	local sub_ver=""
	local str= ""
	print (version)
	for k,v in string.gmatch(version, "(.+)_(.+)")  do
		main_ver = k 
		sub_ver = v 
	end
	print ("main version:" .. main_ver .. ", sub version:" .. sub_ver)
	for substring in main_ver:gmatch("[^.]+")  do 
		table.insert (t,substring)
	end
	for substring in sub_ver:gmatch("[^.]+")  do 
		table.insert (t,substring)
	end
	if #t ~= 7 then
		print ("version need 7 column number")
		os.exit()
	end 
	for key, substring in pairs(t) do 
		byte = tonumber (substring)
		print (byte)
		if byte > 255 then
			print ("version number must lower than 255 for each number")
			os.exit ()
		end

		str = str .. string.char(byte)

	end

	return str
end 

function number2sting(n,byte)
	local str="" 
	for i=3,0,-1 do 
		str = str .. string.char (n >> (8*i) & 0xFF )
	end 
	return str
end
function main ()
	print("firmware_path: " .. firmware_path ..", version: " .. version ..",board_id:".. board_id)
	new_firmware_path =  firmware_path .. ".chk"
	file = io.open (firmware_path,"rb")
	if not file then
		print('Failed to open the file')
		return 
	end
	local f_buffer = file:read("*a")
	file:close()
	header_buf = "*#$^"
	--header_len =  40 +string.len(board_id)
	header_len =  40 + #board_id
	header_len_string = string.char (header_len >> ( 8*3 ) & 0xFF , header_len >> 8*2 & 0xFF ,header_len >> 8 * 1 & 0xFF  , header_len >> 8*0 & 0xFF)
	header_buf = header_buf .. header_len_string
	header_buf = header_buf .. string.char (sku)
	version_buf= validation_version (version)
	header_buf = header_buf .. version_buf
	header_buf=header_buf .. number2sting (calculate_cksum (f_buffer,#f_buffer),4) 
	header_buf=header_buf .. number2sting (rootfs_cksum,4) 
	header_buf=header_buf .. number2sting (#f_buffer,4) 
	
	header_buf=header_buf .. number2sting (rootfs_length,4)
	image_cksum=kernel_cksum
	--header_buf=header_buf .. number2sting (image_cksum,4)
	header_buf=header_buf .. number2sting (calculate_cksum (f_buffer,#f_buffer),4) 
	
	header_buf_tmp = header_buf
	header_buf=header_buf .. number2sting (header_cksum,4)
	header_buf = header_buf .. board_id
	header_cksum = number2sting (calculate_cksum (header_buf,#header_buf),4)
	header_buf_tmp = header_buf_tmp.. header_cksum
	header_buf_tmp = header_buf_tmp .. board_id 
	header_buf = header_buf_tmp

	file = io.open(new_firmware_path,"wb")

	file:write(header_buf)
	file:write(f_buffer)
	file:close()

end
if  #arg ~= 3  then
	
	print ("Must include 4 parameter <firmware_path> <version> <board_id>")
	os.exit()
end

main ()
